src/caffe/proto/caffe.proto

```diff
@@ -347,6 +347,9 @@ message LayerParameter {
   // Parameters shared by loss layers.
   optional LossParameter loss_param = 101;
 
+  // Specify name of source class
+  optional string source_class = 143;
```

Comment:
This pull request is motivated by making Caffe work on Spark (via CaffeOnSpark). source_class was used by CaffeOnSpark to decide which particular Scala class to use for handling a given data source. See [CaffeOnSpark code](https://github.com/yahoo/CaffeOnSpark/blob/master/caffe-grid/src/main/scala/com/yahoo/ml/caffe/DataSource.scala#L114). 

Our original idea is to support custom sources for data layers beyond MemoryDataLayer. For example, to support NLP use cases, we have introduced a VWDataLayer. That's why we propose an optional field at LayerParameter. By having source_class at LayerParemeter, we will enable CaffeOnSpark users to introduce customized parser for different data layers.

FYI, VWDataParameter is defined as below:
message VWDataParameter {
  optional string source = 1;
  optional uint32 batch_size = 2;
  optional bool shuffle = 3 [default = false];
  message TopLayer{
    optional string name = 1;
    optional string type = 2;
    optional uint32 channels = 3;
    optional uint32 height = 4;
    optional uint32 width = 5;
    optional string vw_namespace = 6;
  }
  repeated TopLayer top = 4;
  optional bool sparse = 5 [default = false];
  optional bool memory_input = 6[default = false];
}